Source/WebCore/dom/Element.cpp

```diff
@@ -4881,11 +4876,24 @@ Vector<RefPtr<WebAnimation>> Element::getAnimations(std::optional<GetAnimationsO
     return animations;
 }
 
-ElementIdentifier Element::createElementIdentifier()
+static WeakHashMap<Element, ElementIdentifier>& elementIdentifiersMap()
+{
+    static MainThreadNeverDestroyed<WeakHashMap<Element, ElementIdentifier>> map;
+    return map;
+}
+
+ElementIdentifier Element::identifier() const
 {
-    ASSERT(!hasNodeFlag(NodeFlag::HasElementIdentifier));
-    setNodeFlag(NodeFlag::HasElementIdentifier);
-    return ElementIdentifier::generate();
+    return elementIdentifiersMap().ensure(*this, [] { return ElementIdentifier::generate(); }).iterator->value;
+}
+
+Element* Element::fromIdentifier(ElementIdentifier identifier)
+{
+    for (auto [element, elementIdentifier] : elementIdentifiersMap()) {
```

Comment:
This finds an element even if it’s not in the document any more, only stops returning the element once its memory is deallocated. The older functions returned elements only if they were still in the document. Do we need to add any checks to keep that consistent?

For example, should `dispatchSystemPreviewActionEvent` still dispatch the event in that case? I am thinking we want to preserve the old semantic where it’s not dependent on the timing of garbage collection and other things that can hold a ref to an element. Could be as simple as checking if the element is connected to the document, or some other kind of check.

I am also wondering if these identifiers should be global or scoped to a document. In the end the question is whether we want these identifiers to have the semantic of a weak pointer where they can track an object after it’s unreachable as long as anything is keeping it alive with a ref, or some other semantic where it’s more about observable lifetimes. I don’t mean to make this unnecessarily complicated.

Reply:
Yes, I realize I slightly changed behavior. I believe Tim wrote most of this code so I plan to ask him today if the behavior change is OK or not. If not, I can always add checks at some call sites to make sure that the Element is still connected to the right document.

Note that the caller in internal was about checking if the Element was alive and is actually more correct after my change. Previously, the Element could have leaked but not be in the document anymore and the tests wouldn't have known.